### ISSUE TITLE
Breaks under Kotlin 1.1.0

### ISSUE THREAD

**Issue by davidsowerby** [drmaas/ratpack-kotlin#8]
This issue does not appear under Kotlin 1.0.6

Using your example:

```
fun main(args: Array<String>) {
  ratpack {
    serverConfig {
      development(true)
      port(9000)
    }
    bindings {
      bindInstance("foobar")
    }
    handlers {
      **get**("test") {
        render("hello " + context.get(String::class.java))
      }
    }
  }
}
```

The 'get' fails with the following: 

> Error:(20, 17) Kotlin: Unresolved reference. None of the following candidates is applicable because of receiver type mismatch: 
@InlineOnly public operator inline fun <@OnlyInputTypes K, V> Map<out String, ???>.get(key: String): ??? defined in kotlin.collections

**Comment by davidsowerby** [comment]
I cannot see a way to fix this - as far as I can work out, Kotlin 1.1.0 is resolving 'handlers' differently.  I've [raised an issue](https://youtrack.jetbrains.com/issue/KT-16617) with JetBrains

**Comment by davidsowerby** [comment]
Kotlin have come back quickly on this issue - it has been classified as 'critical' and targeted at 1.1.1

So sticking with 1.0.6 seems like the best option until the fix is released

**Comment by drmaas** [comment]
Good find! I've worked around the issue in the latest commit.

https://github.com/drmaas/ratpack-kotlin/commit/6d2d571698bef87c3dddb57454572ec998512477

 When 1.1.1 is released I should be able to roll back the stuff I had to add.

**Comment by davidsowerby** [comment]
Great, that gives me an option if I want to go to 1.1.0 - for the moment I will stick with 1.0.6 - though coroutines do look interesting ...

**Comment by drmaas** [comment]
Made a release: https://github.com/drmaas/ratpack-kotlin/releases/tag/v0.6.0-rc.1

Should be available on maven central shortly.

**Comment by davidsowerby** [comment]
Looks like you found a better alternative to that proposed by JetBrains - you have kept the 'handlers' name.  Am I reading this correctly?  For the example I raised, you have changed:

> fun handlers(cb: KChain.() -> Unit) = delegate.handlers { KChain(it).cb()

for a more explicit

> fun handlers(cb: KChain.() -> Unit) : KServerSpec = KServerSpec(delegate.handlers { KChain(it).cb() })

thus reducing / removing the need for inference?  I'm guessing this is also what a lot of the other changes are also doing?

**Comment by drmaas** [comment]
Yes, that was actually a fix so that the handlers section could be followed by some other method (by returning KServerSpec instead of null).

The core of the issue seemed to be a problem with their delegation. I had to remove the 'by delegate' class keywords on KChain.kt so that it became a 'dumb' delegate with no auto-delegating to methods in Chain.java. That's why a bunch of new methods were added to KChain.kt, and method signatures changes in KChainAction.kt.

Source compatibility is an interesting option, but I am not sure of the implications of that.

Renaming the methods was not an option for me :)

**Comment by davidsowerby** [comment]
Seems a shame to lose delegation - that is one of my preferred features of Kotlin.  It does seem that they have some problems with name resolution that they have solved differently in 1.1.  I just wonder where else the issue may come up.

The source compatibility options seems more like a short term fix to me.

**Comment by drmaas** [comment]
Agreed about delegation and source compatibility. Will be interesting to see how the language evolves.

**Comment by davidsowerby** [comment]
Just tried your latest version (0.6.0-rc.1), and am getting errors with `path` and `all` with a Handler class, under Kotlin 1.1.0 and 1.0.6

```
all(RootHandler::class.java)
path(RootHandler::class.java)
```

They fail with the IDE showing:


> None of the following functions can be called with the arguments supplied.
all(Class<Handler>) defined in ratpack.kotlin.handling.KChain
all(Handler) defined in ratpack.kotlin.handling.KChain
all(KContext.() → Unit) defined in ratpack.kotlin.handling.KChain

**Comment by drmaas** [comment]
For now you can do `all(RootHandler())` until it gets resolved.

I am not sure why the compiler is not able to figure this out.

**Comment by drmaas** [comment]
Figured it out, opening a new issue to track.

**Comment by drmaas** [comment]
See #9

**Comment by drmaas** [comment]
0.6.0-rc.2 should be available for d/l in a few hours.

**Comment by drmaas** [comment]
@davidsowerby are you good with closing this off?

**Comment by davidsowerby** [comment]
Ah, I haven't tried rc2 yet - it takes a while to appear in Maven Central and I started doing something else!

Give me a day or two, and I'll take a look

**Comment by davidsowerby** [comment]
Confirmed fixed using kotlin your version 0.6.0
